### ISSUE TITLE
[bug] Inconsistent behavior on object fields assignement

### ISSUE THREAD

**Issue by SkeLLLa** [node-config/node-config#514]
### Env
node-config version: ^3.0.0

### Actual behavior:

config/default.js
```js
module.exports = {
  test: {
     a: {}
  }
}
```

test.js
```
const config = require('config');
const test = config.get('test');
test.b = test.b || {}; // b is silently dropped
test.a.b = test.a.b || {}; // b is added

// result test = { a: { b: {}} }
```

### Expected behavior
Ideally it should work as it worked in 2.X.X
```
const config = require('config');
const test = config.get('test');
test.b = test.b || {}; // b is added
test.a.b = test.a.b || {}; // b is added

// result test = { a: { b: {}}, b: {} }
```

Either it should throw an error when such assignment is done.

Or if all config objects in later versions should not be "extensible" then, `test.a.b = test.a.b || {}` should also drop adding `b`.

But better option is to introduce something like env variable switch like `CONFIG_ENABLE_ODD_OBJECT_BEHAVIOR` and control such extensibility.

The line that casing this problem:
```
// Ensure new properties can not be added.
Object.preventExtensions(object)
```
which was added in https://github.com/lorenwest/node-config/pull/507

**Comment by lorenwest** [comment]
The intended behavior is to not allow config mutation after `require('config')`.

Recently @fgheorghe added code to seal the object, and due to this causing potential backward compatibility issues, it was published on a major version 3.0.

It would be better if, in your example, `test.b` wasn't silently dropped and `test.a.b` were sealed as well. Maybe @fgheorghe could weigh in on the possibility of adding that to object sealing.

In your case, if you wish to retain this type of mutability after `require('config')`, the best option is to fix your version at `2.x.x`.

**Comment by SkeLLLa** [comment]
Basically I just want to have some consistent behavior. So if objects sealed - then they should be sealed in all cases and throw error on mutation no matter in what way it's done. If it's not possible, then objects shouldn't be sealed.

**Comment by markstos** [comment]
Co-maintainer here. I agree that we should throw an error on attempted modification and not silently fail.

**Comment by fgheorghe** [comment]
Makes sense, I will investigate the issue further and get back to you with a proposal on how we can throw exceptions.

**Comment by markstos** [comment]
Thanks, @fgheorghe !

**Comment by fgheorghe** [comment]
Apologies for the delay, work and all got in the way. 

I made the following change:
https://github.com/lorenwest/node-config/pull/516/commits/36dd523fcdc73d644808987728f4202f54efdf39

It will properly capture an attempt to mutate child properties, and throw an exception when *new* properties are added and when *existing* properties are updated.

Furthermore, it will raise awareness of the ALLOW_CONFIG_MUTATIONS environment variable, which allows for mutating properties. Basically CONFIG_ENABLE_ODD_OBJECT_BEHAVIOR is already in place using ALLOW_CONFIG_MUTATIONS.

Capturing mutations is done using a Proxy for an object configuration property. Extending array type configuration properties will throw a regular TypeError.

**Comment by jasonwilliams** [comment]
@lorenwest do you know the best way to stub `config` calls now? 
Our code is like this.
```
      sandbox.stub(config, 'get').callsFake((...args) => {
        if (args[0] === 'environment') {
          return 'live';
        }
      });
```
but since moving to version 3 our output is
`TypeError: Cannot define property get, object is not extensible`

**Comment by SkeLLLa** [comment]
@jasonwilliams You can try to using `process.env.ALLOW_CONFIG_MUTATIONS = 'true';` or setting such env variable to make config work as usual.

**Comment by jasonwilliams** [comment]
@SkeLLLa thanks!

**Comment by jdmarshall** [comment]
There is a nasty feature of ES where Trying to update a field set read-only by setProperty will fail silently, and we might not be able to override it or have to do something weird with set() in order to throw the error.

**Comment by jdmarshall** [comment]
I think we are seeing different behavior without strict being enabled in the test code or config.js.

I just ran into #815 which I believe is related here.

**Comment by jdmarshall** [comment]
@SkeLLLa Your silent failure is caused by not using strict mode for the example.

Some of our tests relied on the same sort of silent failure. They broke when I added 'use strict';
